CMakeLists.txt

```diff
@@ -110,9 +110,14 @@ option(USE_LITE_PROTO "Use lite protobuf instead of full." OFF)
 option(USE_LMDB "Use LMDB" OFF)
 option(USE_METAL "Use Metal for iOS build" ON)
 option(USE_NATIVE_ARCH "Use -march=native" OFF)
+if (USE_CUDA OR USE_ROCM)
+  set(USE_GPU TRUE)
+else()
+  set(USE_GPU FALSE)
+endif()
 cmake_dependent_option(
     USE_NCCL "Use NCCL" ON
-    "USE_CUDA;UNIX;NOT APPLE" OFF)
+    "USE_GPU;UNIX;NOT APPLE" OFF)
 cmake_dependent_option(
```

Comment:
@xuhdev Another CMake nuance: when this line was using `USE_CUDA`, would it use the final value of USE_CUDA after all the cmake commands had been processed (as opposed to using the value of `USE_CUDA` set to `ON` just a few lines above in the file)? My theory about why the clang_tidy check fails is that: when we changed it to use `USE_GPU`, that variable is set to TRUE based on the initial value of `USE_CUDA` (and is not updated later), but `USE_CUDA` is set to false in `cmake/Dependencies.cmake:843` when it cannot find a CUDA installation. Do we really want `USE_CUDA` to have a default value of `ON` in `CMakeLists.txt:98`?

Reply:
It uses the value when this line is executed. In this context, it uses the value of `USE_CUDA` a few lines above it.

IMO, whether `USE_CUDA` defaults to ON or OFF is a matter of user experience and shouldn't break builds. One thought to fix this is that, instead of silently turning off `USE_CUDA` when CUDA is not found, we emit an error and ask the user to explicitly turn it off. This should also resolve your issue. What do you think @ezyang ?

Reply:
That feels a little bit like tail wagging dog. If CUDA isn't available on my system, I should do the right thing and just build you a CPU only thing (which implies, I guess, that we shouldn't assume `USE_CUDA` means we are actually building with CUDA?). But I'm not sure if that is idiomatic cmake.

Reply:
@ezyang Either way has its own advantage (explicitly force user to know that CUDA is not available vs. more convenience), but I feel like in this particular case, since CUDA is an essential (and optional) component of PyTorch, it is better to force user to stop unless they acknowledge that CUDA will not be used. If we go the more convenient way, we then might wanna move CUDA detection code upfront so that the other option code knows whether CUDA or GPU is available.

Reply:
I suppose we can give it a try, if this helps solve a bug that is otherwise difficult to solve.

Reply:
@xuhdev I'd like to move forward on this PR before too many conflicts arise from upstream master. It sounds like the consensus is to error out if `USE_CUDA` is set but CUDA is not found. I'd suggest making that change in a separate PR since it's related to this PR, but independent as well. What do you think?

Reply:
Yes, it makes sense to me

Reply:
@xuhdev Can I request you to make that change? I attempted it but `USE_CUDA` is overridden in a few other places before it gets to `cmake/Dependencies.cmake`, which is where I was trying to change the warning to a fatal error. As a result, I cannot test it reliably.

Reply:
@jithunnair-amd Sure, PR opened at  #24939